packages/driver/src/cypress/keyboard.coffee

```diff
@@ -211,6 +213,32 @@ $Keyboard = {
       options.setKey    = "{downarrow}"
       @ensureKey el, null, options, ->
         $selection.moveCursorDown(el)
+
+    ## charCode = 36
+    ## no keyPress
+    ## no textInput
+    ## no input
+    "{home}": (el, options) ->
+      options.charCode  = 36
+      options.keypress  = false
+      options.textInput = false
+      options.input     = false
+      options.setKey    = "{home}"
+      @ensureKey el, null, options, ->
+        $selection.moveCursorToStart(el)
+
+    ## charCode = 35
+    ## no keyPress
+    ## no textInput
+    ## no input
+    "{end}": (el, options) ->
+      options.charCode  = 35
+      options.keypress  = false
+      options.textInput = false
+      options.input     = false
+      options.setKey    = "{end}"
+      @ensureKey el, null, options, ->
+        $selection.moveCursorToEnd(el)
```

Comment:
if not an input, set scrollTop=el.scrollHeight on innermost scrollable div

Reply:
a bit confusing for me.

i know that when you use Home and End, and you are not in input/textarea, it should scroll to top/bottom of page.

however there is a similar story with Up Arrow and Down Arrow, they scroll up/down a little bit.

that being said, I don't see any scrolling stuff in their handlers here. it's the reason I did not add anything to my handlers either.

what am i missing?

Reply:
I see what you're saying about up/down arrow not currently scrolling the page. We probably wouldn't end up fulfilling that default action, and just stick to default action inside of text editables

However, the selection.modify API isn't available in IE11 so we should be hesitant to use it. @silviuavram is your app preventing default on the home key press, or do you need the default action?

Reply:
Resumed working on this and renamed a variable as it was used for two reasons in the same function: `start`. This is what happens when you are not paying attention...

Anyway, about the {home} and {end} keys, my use is to create e2e tests for a Dropdown, and the scenario tests that, with focus on Edit Text and Menu open, Home highlights the first option in the menu, while End highlights the last one.

I want to fully understand how this works in `keyboard.coffee`. If there is an element focused, we call, for example, `$selection.moveCursorToLineStart(el)`, to simulate the cursor behaviour for Home/End inside an `<input>` or `<textarea>`? If no focused element, what happens then? Are the Home/End default behaviours happening anyway? (scrolling to the beginning/end of the page, if no input element is focused).

Or maybe point me to some documentation maybe that will be easier. And also since I mentioned pointing at stuff, I cannot find where to add unit tests for my changes. Can you also help me with this, please?

Thank you for your time! @Bkucera

Reply:
I started to add some unit tests. Will continue tomorrow. You can take a fast look to see if they're on the right track.

I also refactored the `_moveCursorToLineStartOrEnd` method and used only `selection.modify` to move that cursor, for everything (contenteditable, textarea, input). I don't think there is any reason to do some `\n` search and move the cursor with `setSelectionRange` after doing some math magic to locate the line start/end, right?

Hopefully if this is on the right track we can merge it soon. Thanks very much for the help! @Bkucera

Reply:
@silviuavram sure, modern browsers support selection.modify so that's fine. 

I believe you are referring to integration tests, not unit?

Reply:
@Bkucera great to hear that.

As for the tests, I think I've added kind of unit tests. You can check now in `type_spec.coffee` where I added for `{home}` and `{end}`. But please let me know if I need to add more and maybe point me to location. The test codebase is pretty massive.

Thank you very much for assistance!